### ISSUE TITLE
General Code Review 2022/08/31

### ISSUE THREAD

**Issue by ebakker1** [timcanby/MoleculeVAE-re-implementation#1]
This is my general code review on ways I think the code could be improved. I'm no expert software engineer, and Tiago and Miguel know much better, but hopefully it's still useful.

I'll structure the review into comments, with each comment being a broad area with some specific examples.

**Comment by ebakker1** [comment]
## Useful Resources
* Miguel is our software engineering expert at recursive and [his saisho repo](https://github.com/recursiveai/saisho/tree/main/inference) is a great example of how to organise code.
* [This script](https://github.com/ebakker1/general_scripts/blob/main/generic_pytorch_script.py) is my example of how I like to organise training scripts. It might be a bit over engineered, especially for what you need (you could strip out all the reloading stuff) but I find it a good starting point.
* As we've discussed before [lucid rains](https://github.com/lucidrains) has very well structured pytorch repos.

**Comment by ebakker1** [comment]
## General Points
- files like `Dicdata.pkl` and `nohup.out` that appear to be binary files. Better not to have these in the repo and add a `gitignore` to avoid adding these files (like [this one](https://github.com/github/gitignore/blob/main/Python.gitignore)).
- Most files have unused imports. This makes things confusing and your IDE should flag the ones you don’t use so they can be easily removed.
- It's good to have docstrings, particularly on headline classes and important functions. Don’t need documentation on every class and function per se but good to have a thorough documentation on the things that are core to the implementation. For example, the `MolecularVAE` class:
https://github.com/timcanby/MoleculeVAE-re-implementation/blob/9ab3946f11920d870493d6d628079029462f5161/MoleculeVAE-reImplementation/model.py#L21-L27
good to link to original implementation and say where the parameters come from, explain any changes you have made etc.
- docstrings for unclear functions: example of `PrePro`:
https://github.com/timcanby/MoleculeVAE-re-implementation/blob/9ab3946f11920d870493d6d628079029462f5161/MoleculeVAE-reImplementation/model.py#L58-L62
assume it’s preprocessing but hard to know what it does.
- values for network parameters should be parameters you pass into the model or `GLOBAL_VARIABLES` : best not to repeat 292 lots of times like here:
https://github.com/timcanby/MoleculeVAE-re-implementation/blob/9ab3946f11920d870493d6d628079029462f5161/MoleculeVAE-reImplementation/model.py#L27-L32
- Adding a [`setup.py`](http://setup.py) (ideally with a `pyproject.toml` like in saisho) will mean you can install the repo. This is useful if generally but especially if you want to use it in colab.
- Lots of implicit global variables in functions - not a good idea. Example of `train` function:
https://github.com/timcanby/MoleculeVAE-re-implementation/blob/9ab3946f11920d870493d6d628079029462f5161/MoleculeVAE-reImplementation/train.py#L40-L49
`train_loader` is assumed to exist as a global variable as it’s not passed in as a parameter.
- loading a pickle file on each iteration of the test set isn't necessary. Can load once at the beginning:
https://github.com/timcanby/MoleculeVAE-re-implementation/blob/9ab3946f11920d870493d6d628079029462f5161/MoleculeVAE-reImplementation/train.py#L64-L66

**Comment by ebakker1** [comment]
## More specific points
- Why the `1e-2` on the sampling?
https://github.com/timcanby/MoleculeVAE-re-implementation/blob/9ab3946f11920d870493d6d628079029462f5161/MoleculeVAE-reImplementation/model.py#L46-L48
- functions in dataloader need doc strings. These are quite data specific and their use won’t be clear they way something common like `forward` is.
- best to keep imports at the top of the file:
https://github.com/timcanby/MoleculeVAE-re-implementation/blob/9ab3946f11920d870493d6d628079029462f5161/MoleculeVAE-reImplementation/train.py#L105-L111
- generally file opening is done with `with` prevents problems if the code fails between open and close. Here you use open and close:
https://github.com/timcanby/MoleculeVAE-re-implementation/blob/9ab3946f11920d870493d6d628079029462f5161/MoleculeVAE-reImplementation/train.py#L64-L67
